src/Servers/Kestrel/Core/src/Internal/Infrastructure/PipeWriterHelpers/ConcurrentPipeWriter.cs

```diff
@@ -0,0 +1,397 @@
+                _bytesBuffered += bytes;
+                _tailMemory = _tailMemory.Slice(bytes);
+                _nonPassThroughAdvancePending = false;
+            }
+        }
+
+        // This is not exposed to end users. Throw so we find out if we ever start calling this.
+        public override void CancelPendingFlush()
+        {
+            // If we wanted, we could propagate IsCanceled when we do multiple flushes in a loop.
+            // If FlushResult.IsCanceled is true with more data pending to flush, we could complete _currentFlushTcs with canceled flush task,
+            // but rekick the FlushAsync loop.
+            throw new NotImplementedException();
+        }
+
+        // This is not exposed to end users. Throw so we find out if we ever start calling this.
+        public override void Complete(Exception exception = null)
+        {
+            // If we wanted, we could store the complete exception or some sentinel exception instance in a field
+            // if a flush was ongoing. We'd just call the inner Complete() method after the flush loop ended.
+
+            // To simply ensure everything gets returned after the PipeWriter is left in some unknown state (say GetMemory() was
+            // called but not Advance(), or there's a flush pending), just call Abort().
+            throw new NotImplementedException();
+        }
+
+        public override ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken = default)
+        {
+            lock (_sync)
+            {
+                if (_isFlushing)
+                {
+                    return new ValueTask<FlushResult>(_currentFlushTcs.Task);
+                }
+
+                if (_bytesBuffered > 0)
+                {
+                    CopyAndReturnSegmentsUnsynchronized();
+                }
+
+                var flushTask = _innerPipeWriter.FlushAsync(cancellationToken);
+
+                if (flushTask.IsCompletedSuccessfully)
+                {
+                    if (_currentFlushTcs != null)
+                    {
+                        CompleteFlushUnsynchronized(flushTask.GetAwaiter().GetResult(), null);
+                    }
+
+                    return flushTask;
+                }
+
+                _isFlushing = true;
+
+                // Use a TCS instead of something resettable so it can be awaited by multiple awaiters.
+                // _currentFlushTcs might already be set if the last flush that completed, completed between a call to GetMemory() and Advance().
+                _currentFlushTcs ??= new TaskCompletionSource<FlushResult>(TaskCreationOptions.RunContinuationsAsynchronously);
+                return FlushAsyncAwaited(flushTask, cancellationToken);
+            }
+        }
+
+        private async ValueTask<FlushResult> FlushAsyncAwaited(ValueTask<FlushResult> flushTask, CancellationToken cancellationToken)
+        {
+            try
+            {
+                var flushResult = await flushTask;
```

Comment:
move this into the beginning of the while loop and remove the second await flushTask at the end of the loop.